rpcs3/Emu/Io/PadHandler.h

```diff
@@ -238,23 +248,277 @@ struct Pad
+	cfg::string ls_right{ this, "Left Stick Right", "" };
+	cfg::string ls_up   { this, "Left Stick Up", "" };
+	cfg::string rs_left { this, "Right Stick Left", "" };
+	cfg::string rs_down { this, "Right Stick Down", "" };
+	cfg::string rs_right{ this, "Right Stick Right", "" };
+	cfg::string rs_up   { this, "Right Stick Up", "" };
+	cfg::string start   { this, "Start", "" };
+	cfg::string select  { this, "Select", "" };
+	cfg::string ps      { this, "PS Button", "" };
+	cfg::string square  { this, "Square", "" };
+	cfg::string cross   { this, "Cross", "" };
+	cfg::string circle  { this, "Circle", "" };
+	cfg::string triangle{ this, "Triangle", "" };
+	cfg::string left    { this, "Left", "" };
+	cfg::string down    { this, "Down", "" };
+	cfg::string right   { this, "Right", "" };
+	cfg::string up      { this, "Up", "" };
+	cfg::string r1      { this, "R1", "" };
+	cfg::string r2      { this, "R2", "" };
+	cfg::string r3      { this, "R3", "" };
+	cfg::string l1      { this, "L1", "" };
+	cfg::string l2      { this, "L2", "" };
+	cfg::string l3      { this, "L3", "" };
+
+	cfg::_int<0, 1000000> lstickdeadzone{ this, "Left Stick Deadzone", 0 };
+	cfg::_int<0, 1000000> rstickdeadzone{ this, "Right Stick Deadzone", 0 };
+	cfg::_int<0, 1000000> ltriggerthreshold{ this, "Left Trigger Threshold", 0 };
+	cfg::_int<0, 1000000> rtriggerthreshold{ this, "Right Trigger Threshold", 0 };
+	cfg::_int<0, 1000000> padsquircling{ this, "Pad Squircling Factor", 0 };
+
+	cfg::_int<0, 255> colorR{ this, "Color Value R", 0 };
+	cfg::_int<0, 255> colorG{ this, "Color Value G", 0 };
+	cfg::_int<0, 255> colorB{ this, "Color Value B", 0 };
+
+	cfg::_bool enable_vibration_motor_large{ this, "Enable Large Vibration Motor", true };
+	cfg::_bool enable_vibration_motor_small{ this, "Enable Small Vibration Motor", true };
+	cfg::_bool switch_vibration_motors{ this, "Switch Vibration Motors", false };
+
+	bool load()
+	{
+		if (fs::file cfg_file{ cfg_name, fs::read })
+		{
+			return from_string(cfg_file.to_string());
+		}
+
+		return false;
+	}
+
+	void save()
+	{
+		fs::file(cfg_name, fs::rewrite).write(to_string());
+	}
+
+	bool exist()
+	{
+		return fs::is_file(cfg_name);
+	}
+};
+
 class PadHandlerBase
 {
 protected:
+	static const u32 MAX_GAMEPADS = 7;
+
+	bool b_has_deadzones = false;
+	bool b_has_rumble = false;
 	bool b_has_config = false;
+	pad_config m_pad_config;
+
+	template <typename T>
+	T lerp(T v0, T v1, T t) {
+		return std::fma(t, v1, std::fma(-t, v0, v0));
+	}
+
+	// Search an unordered map for a string value and return found keycode
+	int FindKeyCode(std::unordered_map<u32, std::string> map, const std::string& name)
+	{
+		for (auto it = map.begin(); it != map.end(); ++it)
+		{
+			if (it->second == name) return it->first;
+		}
+		return -1;
+	};
+	int FindKeyCode(std::unordered_map<u64, std::string> map, const std::string& name)
+	{
+		for (auto it = map.begin(); it != map.end(); ++it)
+		{
+			if (it->second == name) return it->first;
```

Comment:
In this case, `it->first` is an `u64`. Would it be ok to return an `u64` (64 bits) whereas the return type of this method is an `int` (32 bits)?

Reply:
would long be the correct value here?

Reply:
but Button has u32 as keycode....
no idea. I' haven't noticed any drawback with normal controllers yet

Reply:
xbox controller works fine at least.